pkg/volume/scaleio/sio_client.go

```diff
@@ -0,0 +1,537 @@
+	WaitForAttachedDevice(token string) (string, error)
+	WaitForDetachedDevice(token string) error
+	GetVolumeRefs(string) (int, error)
+}
+
+type sioClient struct {
+	client           *sio.Client
+	gateway          string
+	username         string
+	password         string
+	insecure         bool
+	certsEnabled     bool
+	system           *siotypes.System
+	sysName          string
+	sysClient        *sio.System
+	protectionDomain *siotypes.ProtectionDomain
+	pdName           string
+	pdClient         *sio.ProtectionDomain
+	storagePool      *siotypes.StoragePool
+	spName           string
+	spClient         *sio.StoragePool
+	provisionMode    string
+	sdcPath          string
+	instanceID       string
+	inited           bool
+	diskRegex        *regexp.Regexp
+	mtx              sync.Mutex
+}
+
+func newSioClient(gateway, username, password string, sslEnabled bool) (*sioClient, error) {
+	client := new(sioClient)
+	client.gateway = gateway
+	client.username = username
+	client.password = password
+	if sslEnabled {
+		client.insecure = false
+		client.certsEnabled = true
+	} else {
+		client.insecure = true
+		client.certsEnabled = false
+	}
+	r, err := regexp.Compile(`^emc-vol-\w*-\w*$`)
+	if err != nil {
+		glog.Error(log("failed to compile regex: %v", err))
+		return nil, err
+	}
+	client.diskRegex = r
+
+	// delay client setup/login until init()
+	return client, nil
+}
+
+// init setups client and authenticate
+func (c *sioClient) init() error {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	if c.inited {
+		return nil
+	}
+	glog.V(4).Infoln(log("initializing scaleio client"))
+	client, err := sio.NewClientWithArgs(c.gateway, "", c.insecure, c.certsEnabled)
+	if err != nil {
+		glog.Error(log("failed to create client: %v", err))
+		return err
+	}
+	c.client = client
+	if _, err = c.client.Authenticate(
+		&sio.ConfigConnect{
+			Endpoint: c.gateway,
+			Version:  "",
+			Username: c.username,
+			Password: c.password},
+	); err != nil {
+		glog.Error(log("client authentication failed: %v", err))
+		return err
+	}
+
+	// retrieve system
+	if c.system, err = c.findSystem(c.sysName); err != nil {
+		glog.Error(log("unable to find system %s: %v", c.sysName, err))
+		return err
+	}
+
+	// retrieve protection domain
+	if c.protectionDomain, err = c.findProtectionDomain(c.pdName); err != nil {
+		glog.Error(log("unable to find protection domain %s: %v", c.protectionDomain, err))
+		return err
+	}
+	// retrieve storage pool
+	if c.storagePool, err = c.findStoragePool(c.spName); err != nil {
+		glog.Error(log("unable to find storage pool %s: %v", c.storagePool, err))
+		return err
+	}
+	c.inited = true
+	return nil
+}
+
+func (c *sioClient) Volumes() ([]*siotypes.Volume, error) {
+	if err := c.init(); err != nil {
+		return nil, err
+	}
+	vols, err := c.getVolumes()
+	if err != nil {
+		glog.Error(log("failed to retrieve volumes: %v", err))
+		return nil, err
+	}
+	return vols, nil
+}
+
+func (c *sioClient) Volume(id string) (*siotypes.Volume, error) {
```

Comment:
Looks like you didn't get around to this, but no big deal: this was a nice to have, not a blocker.